### PR TITLE
cache loadUser if not exists

### PR DESCRIPTION
**Pull request by MorrisJobke**
* downstream of https://github.com/owncloud/core/pull/27068

I can't see the requests in our code. Before and after are the same 8 queries run for doing the login (I send the login form and then profiled this via blackfire.)

@icewind1991 @rullzer @blizzz @nickvergessen I would like your feedback on this. We can also go for the stuff that looks good (like the `if` instead the `while`, the PHPDoc comment and the additional unit tests).

### CODE REVIEW

File: lib/private/User/Database.php [hunk 1]
```diff
@@ -88,15 +88,21 @@ public function __construct($eventDispatcher = null) {
 	 * itself, not in its subclasses.
 	 */
 	public function createUser($uid, $password) {
+		unset($this->cache[$uid]); // make sure we are reading from the db
```
Comment (rullzer) [inline lib/private/User/Database.php]: So we don't believe the cache?

File: lib/private/User/Database.php [hunk 2]
```diff
@@ -88,15 +88,21 @@ public function __construct($eventDispatcher = null) {
...
 		if (!$this->userExists($uid)) {
 			$event = new GenericEvent($password);
 			$this->eventDispatcher->dispatch('OCP\PasswordPolicy::validate', $event);
+			unset($this->cache[$uid]); // invalidate non existing user in cache
```
Comment (rullzer) [inline lib/private/User/Database.php]: Lets unset again?

File: lib/private/User/Database.php [hunk 3]
```diff
@@ -88,15 +88,21 @@ public function __construct($eventDispatcher = null) {
...
 			$query = \OC_DB::prepare('INSERT INTO `*PREFIX*users` ( `uid`, `password` ) VALUES( ?, ? )');
 			$result = $query->execute(array($uid, \OC::$server->getHasher()->hash($password)));
 
 			// Clear cache
 			unset($this->cache[$uid]);
 
+			if ($result) {
+				unset($this->cache[$uid]); // invalidate non existing user in cache
```
Comment (rullzer) [inline lib/private/User/Database.php]: And again?

File: lib/private/User/Database.php [hunk 4]
```diff
@@ -222,6 +228,7 @@ public function checkPassword($uid, $password) {
 			if(\OC::$server->getHasher()->verify($password, $storedHash, $newHash)) {
 				if(!empty($newHash)) {
 					$this->setPassword($uid, $password);
+					unset($this->cache[$uid]); // invalidate cache
```
Comment (rullzer) [inline lib/private/User/Database.php]: Why invalidate the cache here?

File: lib/private/User/Database.php [hunk 5]
```diff
@@ -234,10 +241,11 @@ public function checkPassword($uid, $password) {
 	/**
 	 * Load an user in the cache
 	 * @param string $uid the username
-	 * @return boolean
+	 * @return boolean true if user was found, false otherwise
 	 */
 	private function loadUser($uid) {
-		if (!isset($this->cache[$uid])) {
+		// if not in cache (false is a valid value)
+		if (!isset($this->cache[$uid]) && $this->cache[$uid] !== false) {
```
Comment (rullzer) [inline lib/private/User/Database.php]: This condition makes no sense.

**Review by rullzer — CHANGES_REQUESTED** [review]
I think we can just drop this downstream.

File: lib/private/User/Database.php [hunk 6]
```diff
@@ -254,10 +262,15 @@ private function loadUser($uid) {
...
 			$this->cache[$uid] = false;
 
-			while ($row = $result->fetchRow()) {
+			// "uid" is primary key, so there can only be a single result
+			if ($row = $result->fetchRow()) {
 				$this->cache[$uid]['uid'] = $row['uid'];
 				$this->cache[$uid]['displayname'] = $row['displayname'];
+			} else {
+				$result->closeCursor();
+				return false;
```
Comment (MorrisJobke) [inline lib/private/User/Database.php]: @rullzer This is then the only change that makes sense (beside adding the unit tests and the PHPDoc), right?

**Comment by MorrisJobke** [comment]
@rullzer I removed all the cache unsets and only kept the good stuff 😉

File: lib/private/User/Database.php (hunk 6 above)
Comment (rullzer) [inline lib/private/User/Database.php]: It looks a bit weird to me. With the two closeCursors so close together etc. But that is style. And the while is indeed not needed.

So yes. This one, phpdoc and the tests make sense to me.

**Review by LukasReschke — APPROVED** [review]

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**